File: 347-top-k-frequent-elements/top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer,Integer> counts = new HashMap<>();
        for(int n : nums){
            counts.put(n,counts.getOrDefault(n,0)+1);
        }

        // configure the heap put number in the heap
        // but sorted by the frequency
        PriorityQueue<Integer> heap = new PriorityQueue<>((n1,n2) -> counts.get(n1) - counts.get(n2));
        for(int n:counts.keySet()){
            heap.add(n);
            if(heap.size()>k){
                heap.remove();
            }
        }
        int[] ans = new int[k];
        for(int i=0; i<k; i++){
            ans[i] = heap.remove();
        }
        return ans;
    }
}

File: 347-top-k-frequent-elements/top-k-frequent-elements.ts
function topKFrequent(nums: number[], k: number): number[] {
    let map = new Map<number, number>();
    for (let el of nums) {
        map.set(el, 1 + map.get(el) || 1);
    }
    const arr = Array.from(map).sort((a,b) => b[1] - a[1])
    let ans = [];
    for(let i = 0; i < k; i++){
        ans.push(arr[i][0]);
    }
    return ans;
};